File: lazy_chameleon/synthesis/adaptive_moe.py
"""AdaptiveMoE — Dynamic expert expansion. Like Agar.io: split to conquer, merge to aggregate."""
from __future__ import annotations
import asyncio, uuid
from dataclasses import dataclass, field
from typing import Callable, Optional

SCALE_MAP = {"easy": 4, "medium": 8, "hard": 16, "extreme": 32, "research": 64}

ROLE_TEMPLATES = {
    "planner":    "You plan approaches and decompose problems.",
    "researcher": "You gather and synthesise information.",
    "coder":      "You write and debug code.",
    "critic":     "You find flaws and challenge assumptions.",
    "verifier":   "You check correctness and validate outputs.",
    "optimizer":  "You improve performance and efficiency.",
    "explainer":  "You make complex things clear and simple.",
    "security":   "You identify vulnerabilities and risks.",
}


@dataclass
class MicroExpert:
    id: str
    role: str
    goal: str
    specialization: str
    memory: list[str] = field(default_factory=list)
    scratchpad: str = ""
    confidence: float = 0.5
    budget_remaining: int = 50
    parent_id: Optional[str] = None
    children_ids: list[str] = field(default_factory=list)
    alive: bool = True
    result: str = ""


@dataclass
class MoEConfig:
    base_experts: int = 4
    max_experts: int = 32
    split_threshold: float = 0.7
    merge_threshold: float = 0.3
    expert_budget: int = 50


class AdaptiveMoE:
    def __init__(self, config: MoEConfig | None = None):
        self.cfg = config or MoEConfig()
        self._experts: dict[str, MicroExpert] = {}

    def spawn_expert(self, role: str, goal: str, specialization: str = "",
                     parent_id: Optional[str] = None, budget: int = 50) -> MicroExpert:
        eid = str(uuid.uuid4())[:8]
        exp = MicroExpert(id=eid, role=role, goal=goal,
                          specialization=specialization or ROLE_TEMPLATES.get(role, ""),
                          budget_remaining=budget, parent_id=parent_id)
        self._experts[eid] = exp
        if parent_id and parent_id in self._experts:
            self._experts[parent_id].children_ids.append(eid)
        return exp

    def split_expert(self, expert: MicroExpert, n_children: int = 2) -> list[MicroExpert]:
        sub_goals = self._decompose_goal(expert.goal, n_children)
        children  = []
        child_budget = expert.budget_remaining // n_children
        for goal in sub_goals:
            child = self.spawn_expert(
                role=expert.role, goal=goal,
                specialization=expert.specialization,
                parent_id=expert.id, budget=child_budget,
            )
            children.append(child)
        expert.alive = False   # parent consumed by split
        return children

    def merge_experts(self, experts: list[MicroExpert]) -> MicroExpert:
        results = [e.result for e in experts if e.result]
        merged_result = "\n\n---\n\n".join(results)
        avg_conf = sum(e.confidence for e in experts) / max(len(experts), 1)
        merged = self.spawn_expert(
            role="merger", goal="Aggregate results from sub-experts",
            budget=20,
        )
        merged.result = merged_result
        merged.confidence = avg_conf
        for e in experts:
            e.alive = False
        return merged

    def scale_to_task(self, task: str, difficulty: str = "medium") -> list[MicroExpert]:
        n = min(SCALE_MAP.get(difficulty, 8), self.cfg.max_experts)
        roles = list(ROLE_TEMPLATES.keys())
        experts = []
        for i in range(n):
            role = roles[i % len(roles)]
            exp  = self.spawn_expert(
                role=role,
                goal=f"[{role}] {task[:100]}",
                budget=self.cfg.expert_budget,
            )
            experts.append(exp)
        return experts

    async def run_expert(self, expert: MicroExpert, task: str,
                         api_fn: Callable) -> str:
        if not expert.alive or expert.budget_remaining <= 0:
            return ""
        system = (f"You are a {expert.role} expert. {expert.specialization}\n"
                  f"Your specific goal: {expert.goal}\n"
                  f"Be concise and focused. Budget: {expert.budget_remaining} tokens.")
        try:
            result = await api_fn(system, task, expert.role)
            expert.result = result
            expert.confidence = 0.7
            expert.budget_remaining = 0
            return result
        except Exception as e:
            expert.result = f"ERROR: {e}"
            expert.alive  = False
            return ""

    def kill_expert(self, expert_id: str):
        if expert_id in self._experts:
            self._experts[expert_id].alive = False

    def get_alive_experts(self) -> list[MicroExpert]:
        return [e for e in self._experts.values() if e.alive]

    def aggregate_results(self, experts: list[MicroExpert]) -> dict:
        alive    = [e for e in experts if e.result]
        if not alive:
            return {"merged": "", "confidence": 0.0, "n_experts": 0}
        by_role: dict[str, list[str]] = {}
        for e in alive:
            by_role.setdefault(e.role, []).append(e.result)
        sections = []
        for role, results in by_role.items():
            sections.append(f"=== {role.upper()} ===\n" + "\n".join(results[:2]))
        merged = "\n\n".join(sections)
        avg_conf = sum(e.confidence for e in alive) / len(alive)
        return {"merged": merged, "confidence": avg_conf, "n_experts": len(alive)}

    def get_expansion_stats(self) -> dict:
        all_exp = list(self._experts.values())
        return {
            "total_spawned": len(all_exp),
            "alive": len([e for e in all_exp if e.alive]),
            "completed": len([e for e in all_exp if e.result]),
            "by_role": {r: len([e for e in all_exp if e.role == r])
                        for r in ROLE_TEMPLATES},
        }

    def _decompose_goal(self, goal: str, n: int) -> list[str]:
        aspects = [
            f"Core implementation: {goal[:60]}",
            f"Edge cases and validation: {goal[:60]}",
            f"Performance and optimisation: {goal[:60]}",
            f"Error handling: {goal[:60]}",
            f"Testing strategy: {goal[:60]}",
            f"Documentation: {goal[:60]}",
        ]
        return aspects[:n]

File: src/chameleon/adaptive-moe.ts
/**
 * Lazy Chameleon Adaptive MoE Engine
 * 
 * Dynamic cellular expert expansion: split to conquer, merge to aggregate.
 * Decomposes tasks into micro-experts, branches them for deep domain coverage,
 * and synthesizes dense, structured parameter datasets.
 */

export type ExpertRole =
  | 'planner'
  | 'coder'
  | 'critic'
  | 'verifier'
  | 'optimizer'
  | 'security'
  | 'explainer'
  | 'researcher'
  | 'merger';

export interface MicroExpert {
  id: string;
  role: ExpertRole;
  goal: string;
  specialization: string;
  confidence: number;
  budgetRemaining: number;
  parentId?: string;
  childrenIds: string[];
  alive: boolean;
  result: string;
}

export interface MoEConfig {
  baseExperts: number;
  maxExperts: number;
  splitThreshold: number;
  mergeThreshold: number;
  expertBudget: number;
}

export const ROLE_TEMPLATES: Record<ExpertRole, string> = {
  planner: 'Deconstruct approaches, isolate sub-problems, and build dependency DAGs.',
  coder: 'Derive concrete data structures, AST schemas, and implementation logic.',
  critic: 'Expose edge-case hazards, race conditions, and hidden assumptions.',
  verifier: 'Establish mathematical invariants, boundary tests, and proof criteria.',
  optimizer: 'Maximize throughput, eliminate redundant allocations, and align caches.',
  security: 'Audit threat surfaces, privilege escalations, and injection vectors.',
  explainer: 'Synthesize complex findings into high-density technical instructions.',
  researcher: 'Extract domain conventions, API realities, and protocol specs.',
  merger: 'Cross-correlate expert findings and synthesize unified datasets.',
};

export const SCALE_MAP: Record<string, number> = {
  flash: 2,
  turbo: 4,
  easy: 4,
  medium: 8,
  hard: 12,
  deep: 16,
  extreme: 24,
  genius: 32,
  god: 32,
};

export class AdaptiveMoE {
  private experts: Map<string, MicroExpert> = new Map();
  private counter = 0;

  constructor(public config: MoEConfig = {
    baseExperts: 4,
    maxExperts: 32,
    splitThreshold: 0.7,
    mergeThreshold: 0.3,
    expertBudget: 50,
  }) {}

  spawnExpert(
    role: ExpertRole,
    goal: string,
    specialization = '',
    parentId?: string,
    budget = 50,
  ): MicroExpert {
    const id = `exp-${++this.counter}`;
    const expert: MicroExpert = {
      id,
      role,
      goal,
      specialization: specialization || ROLE_TEMPLATES[role] || '',
      confidence: 0.5,
      budgetRemaining: budget,
      parentId,
      childrenIds: [],
      alive: true,
      result: '',
    };
    this.experts.set(id, expert);
    if (parentId && this.experts.has(parentId)) {
      this.experts.get(parentId)!.childrenIds.push(id);
    }
    return expert;
  }

  splitExpert(expert: MicroExpert, nChildren = 2): MicroExpert[] {
    const subGoals = this.decomposeGoal(expert.goal, nChildren);
    const children: MicroExpert[] = [];
    const childBudget = Math.max(10, Math.floor(expert.budgetRemaining / nChildren));

    for (const goal of subGoals) {
      const child = this.spawnExpert(
        expert.role,
        goal,
        expert.specialization,
        expert.id,
        childBudget,
      );
      children.push(child);
    }
    expert.alive = false; // Parent cell consumed by division
    return children;
  }

  mergeExperts(experts: MicroExpert[]): MicroExpert {
    const results = experts.filter((e) => e.result).map((e) => e.result);
    const mergedResult = results.join('\n\n---\n\n');
    const avgConf = experts.reduce((sum, e) => sum + e.confidence, 0) / Math.max(experts.length, 1);

    const merged = this.spawnExpert('merger', 'Aggregate results from sub-experts', '', undefined, 20);
    merged.result = mergedResult;
    merged.confidence = avgConf;

    for (const e of experts) {
      e.alive = false;
    }
    return merged;
  }

  scaleToTask(task: string, difficulty = 'medium'): MicroExpert[] {
    const count = Math.min(SCALE_MAP[difficulty] || 8, this.config.maxExperts);
    const roles: ExpertRole[] = ['planner', 'coder', 'critic', 'verifier', 'optimizer', 'security'];
    const spawned: MicroExpert[] = [];

    for (let i = 0; i < count; i++) {
      const role = roles[i % roles.length];
      const exp = this.spawnExpert(role, `[${role}] ${task.slice(0, 120)}`, '', undefined, this.config.expertBudget);
      spawned.push(exp);
    }
    return spawned;
  }

  aggregateResults(experts: MicroExpert[]): {
    merged: string;
    confidence: number;
    expertCount: number;
    byRole: Record<string, string[]>;
  } {
    const completed = experts.filter((e) => e.result);
    const byRole: Record<string, string[]> = {};

    for (const e of completed) {
      if (!byRole[e.role]) byRole[e.role] = [];
      byRole[e.role].push(e.result);
    }

    const sections: string[] = [];
    for (const [role, results] of Object.entries(byRole)) {
      sections.push(`=== MOE CELL [${role.toUpperCase()}] ===\n` + results.join('\n\n'));
    }

    const avgConf = completed.length > 0
      ? completed.reduce((sum, e) => sum + e.confidence, 0) / completed.length
      : 0;

    return {
      merged: sections.join('\n\n'),
      confidence: Number(avgConf.toFixed(2)),
      expertCount: completed.length,
      byRole,
    };
  }

  private decomposeGoal(goal: string, n: number): string[] {
    const aspects = [
      `Core structural logic: ${goal}`,
      `Edge cases and boundary invariant proof: ${goal}`,
      `Performance, caching, and allocation optimization: ${goal}`,
      `Concurrency, atomicity, and error handling: ${goal}`,
      `Automated testing and verification harness: ${goal}`,
    ];
    return aspects.slice(0, n);
  }
}
